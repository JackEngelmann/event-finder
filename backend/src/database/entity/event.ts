import { Connection, Entity, Column, PrimaryGeneratedColumn } from 'typeorm'

@Entity('event')
export class EventDataModel {
    @Column({ type: 'decimal', nullable: true })
    admissionFee?: number | null

    @Column({ type: 'decimal', nullable: true })
    admissionFeeWithDiscount?: number | null

    @Column({ type: 'int', nullable: true })
    amountOfFloors?: number | null

    @Column()
    clubId!: number

    @Column()
    date!: string

    @Column({
        type: 'text',
        nullable: true,
        charset: 'utf8mb4',
        collation: 'utf8mb4_general_ci',
    })
    description?: string | null

    @PrimaryGeneratedColumn()
    id!: number

    @Column({ type: 'text', nullable: true })
    imageUrl?: string | null

    @Column({ type: 'text', nullable: true })
    link?: string | null

    @Column({ type: 'int', nullable: true })
    minimumAge?: number | null

    @Column()
    name!: string

    @Column({ type: 'int', nullable: true })
    priceCategory?: number | null

    @Column({
        type: 'text',
        nullable: true,
        charset: 'utf8mb4',
        collation: 'utf8mb4_general_ci',
    })
    special?: string | null
}

export class EventModel {
    private connection: Connection

    constructor(connection: Connection) {
        this.connection = connection
    }

    async createEvent(input: {
        admissionFee?: number
        admissionFeeWithDiscount?: number
        amountOfFloors?: number
        clubId: number
        date: string
        description?: String
        genreIds?: number[]
        imageUrl?: string
        link?: string
        minimumAge?: number
        name: string
        priceCategory?: number
        special?: string
    }) {
        const event = new EventDataModel()
        Object.assign(event, input)
        await this.connection.manager.save(event)

        return event.id
    }

    async deleteEvent(id: number) {
        const event = await this.connection.manager.findOneOrFail(
            EventDataModel,
            id
        )
        await this.connection.manager.remove(event)
    }

    async getEvent(id: number) {
        return await this.connection.manager.findOne(EventDataModel, id)
    }

    async getEvents() {
        return await this.connection.manager.find(EventDataModel)
    }

    async getEventsFromClub(clubId: number) {
        return await this.connection.manager.find(EventDataModel, { clubId })
    }

    async updateEvent(input: {
        admissionFee?: number
        admissionFeeWithDiscount?: number
        amountOfFloors?: number
        clubId: number
        date: string
        description?: string
        genreIds?: number[]
        id: number
        imageUrl?: string
        link?: string
        minimumAge?: number
        name: string
        priceCategory?: number
        special?: string
    }) {
        const event = await this.connection.manager.findOne(
            EventDataModel,
            input.id
        )
        if (!event) return undefined

        event.admissionFee = input.admissionFee || null
        event.admissionFeeWithDiscount = input.admissionFeeWithDiscount || null
        event.amountOfFloors = input.amountOfFloors || null
        event.clubId = input.clubId
        event.date = input.date
        event.description = input.description || null
        event.imageUrl = input.imageUrl || null
        event.link = input.link || null
        event.minimumAge = input.minimumAge || null
        event.name = input.name
        event.priceCategory = input.priceCategory || null
        event.special = input.special || null

        await this.connection.manager.save(event)
    }
}
