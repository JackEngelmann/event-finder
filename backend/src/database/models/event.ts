import { Database } from '../database'

export class EventDataModel {
    admissionFee?: number
    admissionFeeWithDiscount?: number
    amountOfFloors?: number
    clubId: string
    date: string
    description?: string
    id: number
    imageUrl?: string
    link?: string
    minimumAge?: number
    name: string
    priceCategory?: 1 | 2 | 3
    special?: string
    constructor(row: any) {
        this.admissionFee = row.admissionfee
        this.admissionFeeWithDiscount = row.admissionfeewithdiscount
        this.amountOfFloors = row.amountoffloors
        this.clubId = row.clubid
        this.date = row.date
        this.description = row.description
        this.id = row.id
        this.imageUrl = row.imageurl
        this.link = row.link
        this.minimumAge = row.minimumage
        this.name = row.name
        this.priceCategory = row.pricecategory
        this.special = row.special
    }
}

export class EventModel {
    private db: Database

    constructor(db: Database) {
        this.db = db
    }

    async createEvent(input: {
        admissionFee?: number
        admissionFeeWithDiscount?: number
        amountOfFloors?: number
        clubId: number
        date: string
        description?: String
        genreIds?: number[]
        imageUrl?: string
        link?: string
        minimumAge?: number
        name: string
        priceCategory?: number
        special?: string
    }) {
        return this.db
            .get(
                `
                INSERT INTO event (
                    admissionFee,
                    admissionFeeWithDiscount,
                    amountOfFloors,
                    clubId,
                    date,
                    description,
                    imageUrl,
                    link,
                    minimumAge,
                    name,
                    priceCategory,
                    special
                ) VALUES (
                    $1,
                    $2,
                    $3,
                    $4,
                    $5,
                    $6,
                    $7,
                    $8,
                    $9,
                    $10,
                    $11,
                    $12
                ) RETURNING Id
            `,
                [
                    input.admissionFee,
                    input.admissionFeeWithDiscount,
                    input.amountOfFloors,
                    input.clubId,
                    input.date,
                    input.description,
                    input.imageUrl,
                    input.link,
                    input.minimumAge,
                    input.name,
                    input.priceCategory,
                    input.special,
                ]
            )
            .then(res => res.id)
            .catch(err => {
                console.error(err)
            })
    }

    deleteEvent(id: number) {
        return this.db.run('DELETE FROM event WHERE id = $1', [id])
    }

    async getEvent(id: number) {
        const sql = 'SELECT * FROM event WHERE id = $1'
        const params = [id]
        const row = await this.db.get(sql, params)
        if (!row) return undefined
        const event = new EventDataModel(row)
        return event
    }

    async getEvents() {
        const sql = 'SELECT * FROM event'
        const rows = await this.db.all(sql, [])
        return rows.map(r => new EventDataModel(r))
    }

    async getEventsFromClub(clubId: number) {
        const sql = 'SELECT * FROM event WHERE clubId = $1'
        const rows = await this.db.all(sql, [clubId])
        return rows.map(r => new EventDataModel(r))
    }

    updateEvent(input: {
        admissionFee?: number
        admissionFeeWithDiscount?: number
        amountOfFloors?: number
        clubId: number
        date: string
        description?: String
        genreIds?: number[]
        id: number
        imageUrl?: string
        link?: string
        minimumAge?: number
        name: string
        priceCategory?: number
        special?: string
    }) {
        return this.db.run(
            `
                UPDATE event 
                SET
                    admissionFee = $1,
                    admissionFeeWithDiscount = $2,
                    amountOfFloors = $3,
                    clubId = $4,
                    date = $5,
                    description = $6,
                    imageUrl = $7,
                    link = $8,
                    minimumAge = $9,
                    name = $10,
                    priceCategory = $11,
                    special = $12
                WHERE id = $13
            `,
            [
                input.admissionFee,
                input.admissionFeeWithDiscount,
                input.amountOfFloors,
                input.clubId,
                input.date,
                input.description,
                input.imageUrl,
                input.link,
                input.minimumAge,
                input.name,
                input.priceCategory,
                input.special,
                input.id,
            ]
        )
    }
}
