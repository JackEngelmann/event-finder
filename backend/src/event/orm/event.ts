import {
    Connection,
    Entity,
    Column,
    PrimaryGeneratedColumn,
    EntityRepository,
    Repository,
} from 'typeorm'

@Entity('event')
export class EventDataModel {
    @Column({ type: 'decimal', nullable: true })
    admissionFee?: number | null

    @Column({ type: 'decimal', nullable: true })
    admissionFeeWithDiscount?: number | null

    @Column({ type: 'int', nullable: true })
    amountOfFloors?: number | null

    @Column()
    clubId!: number

    @Column()
    date!: string

    @Column({
        type: 'text',
        nullable: true,
        charset: 'utf8mb4',
        collation: 'utf8mb4_general_ci',
    })
    description?: string | null

    @PrimaryGeneratedColumn()
    id!: number

    // LEGACY
    @Column({ type: 'text', nullable: true })
    link?: string | null

    @Column({ type: 'int', nullable: true })
    minimumAge?: number | null

    @Column()
    name!: string

    @Column({ type: 'int', nullable: true })
    priceCategory?: number | null

    @Column({
        type: 'text',
        nullable: true,
        charset: 'utf8mb4',
        collation: 'utf8mb4_general_ci',
    })
    special?: string | null
}

@EntityRepository(EventDataModel)
export class EventRepository extends Repository<EventDataModel> {
    async createAndSave(input: {
        admissionFee?: number
        admissionFeeWithDiscount?: number
        amountOfFloors?: number
        clubId: number
        date: string
        description?: string
        genreIds?: number[]
        minimumAge?: number
        name: string
        priceCategory?: number
        special?: string
    }) {
        const event = this.create(input)
        return await this.save(event)
    }

    async findEventsForClub(clubId: number) {
        return await this.find({ clubId })
    }
}
