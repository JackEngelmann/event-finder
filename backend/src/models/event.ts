import { Database } from 'sqlite3'

export class EventDataModel {
    admissionFee?: number
    admissionFeeWithDiscount?: number
    amountOfFloors?: number
    clubId: string
    date: string
    description?: string
    id: number
    imageUrl?: string
    minimumAge?: number
    name: string
    priceCategory?: 1 | 2 | 3
    special?: string
    constructor(row: any) {
        this.admissionFee = row.admissionFee
        this.admissionFeeWithDiscount = row.admissionFeeWithDiscount
        this.amountOfFloors = row.amountOfFloors
        this.clubId = row.clubId
        this.date = row.date
        this.description = row.description
        this.id = row.id
        this.imageUrl = row.imageUrl
        this.minimumAge = row.minimumAge
        this.name = row.name
        this.priceCategory = row.priceCategory
        this.special = row.special
    }
}

export class EventModel {
    private db: Database

    constructor(db: Database) {
        this.db = db
    }

    createEvent(input: {
        name: string
        description?: String
        date: string
        clubId: number
        genreIds?: number[]
        special?: string
        priceCategory?: number
        admissionFee?: number
        admissionFeeWithDiscount?: number
        minimumAge?: number
        amountOfFloors?: number
    }) {
        return new Promise<number>((resolve, reject) => {
            this.db.run(
                `
                INSERT INTO event (
                    name,
                    date,
                    description,
                    clubId,
                    special,
                    priceCategory,
                    admissionFee,
                    admissionFeeWithDiscount,
                    minimumAge,
                    amountOfFloors
                ) VALUES (
                    $name,
                    $date,
                    $description,
                    $clubId,
                    $special,
                    $priceCategory,
                    $admissionFee,
                    $admissionFeeWithDiscount,
                    $minimumAge,
                    $amountOfFloors
                )
            `,
                {
                    $name: input.name,
                    $date: input.date,
                    $description: input.description,
                    $clubId: input.clubId,
                    $special: input.special,
                    $priceCategory: input.priceCategory,
                    $admissionFee: input.admissionFee,
                    $admissionFeeWithDiscount: input.admissionFeeWithDiscount,
                    $minimumAge: input.minimumAge,
                    $amountOfFloors: input.amountOfFloors,
                },
                function(err) {
                    if (err) reject(err)
                    resolve(this.lastID)
                }
            )
        })
    }

    deleteEvent(id: number) {
        return new Promise<void>((resolve, reject) => {
            this.db.run('DELETE FROM event WHERE id = $id', { $id: id }, err =>
                err ? reject(err) : resolve()
            )
        })
    }

    getEvent(id: number) {
        const sql = 'SELECT * FROM event WHERE id = ?'
        const params = [id]
        return new Promise<EventDataModel | undefined>((resolve, reject) => {
            this.db.get(sql, params, async (err, row) => {
                if (err) return reject(err)
                if (!row) {
                    return resolve(undefined)
                }
                const event = new EventDataModel(row)
                resolve(event)
            })
        })
    }

    getEvents() {
        const sql = 'SELECT * FROM event'
        return new Promise<EventDataModel[]>((resolve, reject) => {
            this.db.all(sql, [], async (err, rows) => {
                if (err) return reject(err)
                const events = rows.map(r => new EventDataModel(r))
                resolve(events)
            })
        })
    }

    getEventsFromClub(clubId: number) {
        const sql = 'SELECT * FROM event WHERE clubId = $clubId'
        return new Promise<EventDataModel[]>((resolve, reject) => {
            this.db.all(sql, { $clubId: clubId }, async (err, rows) => {
                if (err) return reject(err)
                const events = rows.map(r => new EventDataModel(r))
                resolve(events)
            })
        })
    }

    updateEvent(input: {
        id: number
        name: string
        description?: String
        date: string
        clubId: number
        genreIds?: number[]
        special?: string
        priceCategory?: number
        admissionFee?: number
        admissionFeeWithDiscount?: number
        minimumAge?: number
        amountOfFloors?: number
    }) {
        return new Promise<number>((resolve, reject) => {
            this.db.run(
                `
                UPDATE event 
                SET
                    admissionFee = $admissionFee,
                    admissionFeeWithDiscount = $admissionFeeWithDiscount,
                    amountOfFloors = $amountOfFloors,
                    clubId = $clubId,
                    date = $date,
                    description = $description,
                    minimumAge = $minimumAge,
                    name = $name,
                    priceCategory = $priceCategory,
                    special = $special
                WHERE id = $id
            `,
                {
                    $admissionFee: input.admissionFee,
                    $admissionFeeWithDiscount: input.admissionFeeWithDiscount,
                    $amountOfFloors: input.amountOfFloors,
                    $clubId: input.clubId,
                    $date: input.date,
                    $description: input.description,
                    $id: input.id,
                    $minimumAge: input.minimumAge,
                    $name: input.name,
                    $priceCategory: input.priceCategory,
                    $special: input.special,
                },
                function(err) {
                    if (err) reject(err)
                    resolve(this.lastID)
                }
            )
        })
    }
}
