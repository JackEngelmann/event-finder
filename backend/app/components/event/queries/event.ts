import { AppContext } from '../../../infrastructure/appContext'
import { EventRepository } from '../orm/event'

export async function queryEvent(appContext: AppContext, id: number) {
    const { db } = appContext
    return await db.getCustomRepository(EventRepository).findOne({ id })
}
