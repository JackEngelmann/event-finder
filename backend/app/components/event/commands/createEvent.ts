import { AppContext } from '../../../infrastructure/appContext'
import { FileUpload } from 'graphql-upload'
import { createLinksForEvent } from '../../link/commands/createLinksForEvent'
import { LinkType } from '../../link/orm/link'
import { EventRepository } from '../orm/event'
import { setGenresForEvent } from '../../genre/commands/setGenresForEvent'
import { setImageUrlsForEvent } from '../../image/commands/setImageUrlsForEvent'
import logger from '../../../infrastructure/logger'

export type CreateEventInput = {
    admissionFee?: number
    admissionFeeWithDiscount?: number
    amountOfFloors?: number
    clubId: number
    date: string
    description?: string
    genreIds?: number[]
    image?: Promise<FileUpload>
    imageUrls?: string[]
    links?: LinkType[]
    minimumAge?: number
    name: string
    priceCategory?: number
    special?: string
}

export async function createEvent(appContext: AppContext, input: CreateEventInput) {
    const { db } = appContext
    try {
        const event = await db
            .getCustomRepository(EventRepository)
            .createAndSave(input)
        await setGenresForEvent(appContext, {
            eventId: event.id,
            genreIds: input.genreIds,
        })
        if (input.imageUrls) {
            await setImageUrlsForEvent(appContext, {
                eventId: event.id,
                imageUrls: input.imageUrls,
            })
        }
        if (input.links) {
            await createLinksForEvent(appContext, {
                eventId: event.id,
                links: input.links,
            })
        }
        return event.id
    } catch (error) {
        logger.error(error)
        throw error
    }
}
