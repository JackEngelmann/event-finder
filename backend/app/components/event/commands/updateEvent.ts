import { AppContext } from '../../../infrastructure/appContext'
import { EventModel } from '../orm/event'
import { EventGenreModel } from '../../genre/orm/eventGenre'
import { EventImageModel } from '../../image/orm/eventImage'
import { LinkType } from '../../link/orm/link'
import { setLinksForEvent } from '../../link/commands/setLinksForEvent'

export type UpdateEventInput = {
    admissionFee?: number
    admissionFeeWithDiscount?: number
    amountOfFloors?: number
    clubId: number
    date: string
    description?: string
    genreIds?: number[]
    id: number
    imageUrls?: string[]
    links?: LinkType[]
    minimumAge?: number
    name: string
    priceCategory?: number
    special?: string
}

export function updateEvent(appContext: AppContext, input: UpdateEventInput) {
    const { db } = appContext
    const eventModel = new EventModel(db)
    const eventGenreModel = new EventGenreModel(db)
    const eventImageModel = new EventImageModel(db)
    return new Promise(async (resolve, reject) => {
        try {
            await eventModel.updateEvent(input)
            await eventGenreModel.setGenresForAnEvent(input.id, input.genreIds)
            await eventImageModel.setImageUrlsForEvent(
                input.id,
                input.imageUrls
            )
            if (input.links) {
                await setLinksForEvent(appContext, {
                    eventId: input.id,
                    links: input.links,
                })
            }
            resolve()
        } catch (err) {
            console.error(err)
            reject(err)
        }
    })
}
