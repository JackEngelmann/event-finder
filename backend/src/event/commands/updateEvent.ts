import { AppContext } from '../../infrastructure/appContext'
import { LinkType } from '../../link/orm/link'
import { setLinksForEvent } from '../../link/commands/setLinksForEvent'
import { EventRepository } from '../orm/event'
import { setGenresForEvent } from '../../genre/commands/setGenresForEvent'
import { setImageUrlsForEvent } from '../../image/commands/setImageUrlsForEvent'
import logger from '../../infrastructure/logger'

export type UpdateEventInput = {
    admissionFee?: number
    admissionFeeWithDiscount?: number
    amountOfFloors?: number
    clubId: number
    date: string
    description?: string
    genreIds?: number[]
    id: number
    imageUrls?: string[]
    links?: LinkType[]
    minimumAge?: number
    name: string
    priceCategory?: number
    special?: string
}

export async function updateEvent(appContext: AppContext, input: UpdateEventInput) {
    const { db } = appContext
    const eventRepository = db.getCustomRepository(EventRepository)
    try {
        await eventRepository.update(input.id, {
            admissionFee: input.admissionFee || null,
            admissionFeeWithDiscount:
                input.admissionFeeWithDiscount || null,
            amountOfFloors: input.amountOfFloors || null,
            clubId: input.clubId,
            date: input.date,
            description: input.description || null,
            id: input.id,
            minimumAge: input.minimumAge || null,
            name: input.name,
            priceCategory: input.priceCategory || null,
            special: input.special || null,
        })
        await setGenresForEvent(appContext, {
            eventId: input.id,
            genreIds: input.genreIds,
        })
        if (input.imageUrls) {
            await setImageUrlsForEvent(appContext, {
                eventId: input.id,
                imageUrls: input.imageUrls,
            })
        }
        if (input.links) {
            await setLinksForEvent(appContext, {
                eventId: input.id,
                links: input.links,
            })
        }
    } catch (error) {
        logger.error(error)
        throw error
    }
}
