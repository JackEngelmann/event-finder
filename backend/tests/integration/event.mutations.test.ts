import {
    eventFragment,
    ApolloHttpTestServer,
    ApolloTestServer,
    createApolloTestServer,
} from '../utils'
import { CreateEventInput } from '../../src/event/commands/createEvent'
import { UpdateEventInput } from '../../src/event/commands/updateEvent'
import { queryEvent } from '../../src/event/queries/event'

const DB_NAME = 'eventmutationdb'
const createEventMutation = `
    mutation CreateEvent($input: CreateEventInput!) {
        createEvent(input: $input) {
            event {
                ${eventFragment}
            }
        }
    }
`
const updateEventMutation = `
    mutation UpdateEvent($input: UpdateEventInput!) {
        updateEvent(input: $input) {
            event {
                ${eventFragment}
            }
        }
    }
`
const deleteEventMutation = `
    mutation DeleteEvent($id: Int!) {
        deleteEvent(id: $id) {
            id
        }
    }
`

let apolloTestServer: ApolloTestServer | undefined

beforeEach(async () => {
    apolloTestServer = await createApolloTestServer({
        isAdmin: false,
        dbName: DB_NAME,
        insertTestData: true,
    })
})

afterEach(async () => {
    if (apolloTestServer) await apolloTestServer.destroy()
    apolloTestServer = undefined
})

describe('event mutations: ', () => {
    describe('create event', () => {
        test('create an event with only required information', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                isAdmin: true,
                dbName: DB_NAME,
                insertTestData: true,
            })
            const input: CreateEventInput = {
                name: 'event-name',
                date: 'event-date',
                clubId: 1,
            }
            // act
            const result = await apolloTestServer.client.mutate({
                mutation: createEventMutation,
                variables: { input },
            })

            // assert
            expect(result.data).toBeDefined()
            expect(result).toMatchSnapshot()
        })

        test('create an event with empty genres', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                isAdmin: true,
                dbName: DB_NAME,
                insertTestData: true,
            })
            const input: CreateEventInput = {
                name: 'event-name',
                date: 'event-date',
                clubId: 1,
                genreIds: [],
            }

            // act
            const result = await apolloTestServer.client.mutate({
                mutation: createEventMutation,
                variables: { input },
            })

            // assert
            expect(result.data).toBeDefined()
            expect(result).toMatchSnapshot()
        })

        test('create an event with all fields', async () => {
            apolloTestServer = await createApolloTestServer({
                isAdmin: true,
                dbName: DB_NAME,
                insertTestData: true,
            })
            const input: CreateEventInput = {
                name: 'event-name',
                description: 'event-description',
                date: 'event-date',
                clubId: 1,
                genreIds: [1, 2],
                special: 'event-special',
                imageUrls: ['image-url-1', 'image-url-2'],
                priceCategory: 2,
                admissionFee: 10.0,
                admissionFeeWithDiscount: 4.0,
                minimumAge: 18,
                links: [{ href: 'link', type: 'FACEBOOK' }],
                amountOfFloors: 3,
            }
            const result = await apolloTestServer.client.mutate({
                mutation: createEventMutation,
                variables: { input },
            })
            expect(result.errors).toBeUndefined()
            expect(result.data).toBeDefined()
            expect(result).toMatchSnapshot()
        })

        test('create an event without permission shoul yield error', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                isAdmin: false,
                dbName: DB_NAME,
                insertTestData: true,
            })
            const input: CreateEventInput = {
                name: 'event-name',
                date: 'event-date',
                clubId: 1,
            }

            // act
            const result = await apolloTestServer.client.mutate({
                mutation: createEventMutation,
                variables: { input },
            })

            // assert
            expect(result.data).toBeNull()
            expect(result.errors).toBeDefined()
            expect(result.errors).toMatchSnapshot()
        })
    })

    describe('update event', () => {
        test('to only have required fields', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                isAdmin: true,
                dbName: DB_NAME,
                insertTestData: true,
            })
            const input: UpdateEventInput = {
                id: 1,
                name: 'updated-name',
                date: 'updated-date',
                clubId: 2,
            }

            // act
            const result = await apolloTestServer.client.mutate({
                mutation: updateEventMutation,
                variables: { input },
            })

            // assert
            expect(result.data).toBeDefined()
            expect(result).toMatchSnapshot()
        })

        test('return null for event when updating a non-existent event', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                isAdmin: true,
                dbName: DB_NAME,
                insertTestData: true,
            })
            const input: UpdateEventInput = {
                id: 10000,
                name: 'updated-name',
                date: 'updated-date',
                clubId: 2,
            }

            // act
            const result = await apolloTestServer.client.mutate({
                mutation: updateEventMutation,
                variables: { input },
            })

            // assert
            expect(result.errors).toBeUndefined()
            expect(result.data).toBeDefined()
            expect(result.data!.updateEvent).toBeDefined()
            expect(result.data!.updateEvent.event).toBeNull()
        })

        test('with empty genre list', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                isAdmin: true,
                dbName: DB_NAME,
                insertTestData: true,
            })
            const input: UpdateEventInput = {
                id: 1,
                name: 'updated-name',
                date: 'updated-date',
                clubId: 2,
                genreIds: [],
            }

            // act
            const result = await apolloTestServer.client.mutate({
                mutation: updateEventMutation,
                variables: { input },
            })

            // assert
            expect(result.data).toBeDefined()
            expect(result).toMatchSnapshot()
        })

        test('with all fields', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                dbName: DB_NAME,
                isAdmin: true,
                insertTestData: true,
            })
            const input: UpdateEventInput = {
                id: 1,
                name: 'updated-name',
                date: 'updated-date',
                clubId: 2,
                description: 'updated-description',
                genreIds: [2, 3],
                imageUrls: ['image-url-1', 'image-url-2'],
                special: 'updated-special',
                priceCategory: 1,
                admissionFee: 20.0,
                admissionFeeWithDiscount: 3.0,
                minimumAge: 16,
                amountOfFloors: 2,
                links: [{ href: 'updated-link', type: 'FACEBOOK' }],
            }

            // act
            const result = await apolloTestServer.client.mutate({
                mutation: updateEventMutation,
                variables: { input },
            })

            // assert
            expect(result.errors).toBeUndefined()
            expect(result.data).toBeDefined()
            expect(result).toMatchSnapshot()
        })

        test('without permission should yield error', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                isAdmin: false,
                dbName: DB_NAME,
                insertTestData: true,
            })
            const input: UpdateEventInput = {
                id: 1,
                name: 'updated-name',
                date: 'updated-date',
                clubId: 2,
            }

            // act
            const result = await apolloTestServer.client.mutate({
                mutation: updateEventMutation,
                variables: { input },
            })

            // assert
            expect(result.data).toBeNull()
            expect(result.errors).toBeDefined()
            expect(result).toMatchSnapshot()
        })

        test('with emojis in description', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                isAdmin: true,
                dbName: DB_NAME,
                insertTestData: true,
            })
            const emojiDescription = '🍃🔝🎉📢👉🌵🎵'
            const input: UpdateEventInput = {
                id: 1,
                name: 'updated-name',
                date: 'updated-date',
                description: emojiDescription,
                clubId: 2,
            }

            // act
            const result = await apolloTestServer.client.mutate({
                mutation: updateEventMutation,
                variables: { input },
            })

            // assert
            expect(result.data).toBeDefined()
            expect(result.data!.updateEvent.event.description).toBe(
                emojiDescription
            )
            expect(result).toMatchSnapshot()
        })
    })

    describe('delte event', () => {
        test('delete an event', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                dbName: DB_NAME,
                isAdmin: true,
                insertTestData: true,
            })
            const { appContext, client } = apolloTestServer
            const id = 1

            const event = await queryEvent(appContext, id)
            expect(event).toBeDefined()

            // act
            await client.mutate({
                mutation: deleteEventMutation,
                variables: { id },
            })
            const deletedEvent = await queryEvent(appContext, id)

            // assert
            expect(deletedEvent).toBeUndefined()
        })

        test('without permission should yield error', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                dbName: DB_NAME,
                isAdmin: false,
                insertTestData: true,
            })
            const { client } = apolloTestServer

            // act
            const result = await client.mutate({
                mutation: deleteEventMutation,
                variables: { id: 1 },
            })

            // assert
            expect(result.data).toBeNull()
            expect(result.errors).toBeDefined()
            expect(result.errors).toMatchSnapshot()
        })
    })
})
