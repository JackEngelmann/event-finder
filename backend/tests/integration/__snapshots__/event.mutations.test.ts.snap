// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`event mutations:  create event create an event with all fields 1`] = `
{
  "data": {
    "createEvent": {
      "event": {
        "admissionFee": 10,
        "admissionFeeWithDiscount": 4,
        "amountOfFloors": 3,
        "club": {
          "id": 1,
          "name": "name",
        },
        "date": "event-date",
        "description": "event-description",
        "genres": [
          {
            "id": 1,
            "name": "genre1",
          },
          {
            "id": 2,
            "name": "genre2",
          },
        ],
        "id": 2,
        "imageUrls": [
          "image-url-1",
          "image-url-2",
        ],
        "links": [
          {
            "href": "link",
            "type": "FACEBOOK",
          },
        ],
        "minimumAge": 18,
        "name": "event-name",
        "priceCategory": 2,
        "special": "event-special",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`event mutations:  create event create an event with empty genres 1`] = `
{
  "data": {
    "createEvent": {
      "event": {
        "admissionFee": null,
        "admissionFeeWithDiscount": null,
        "amountOfFloors": null,
        "club": {
          "id": 1,
          "name": "name",
        },
        "date": "event-date",
        "description": null,
        "genres": [],
        "id": 2,
        "imageUrls": [],
        "links": [],
        "minimumAge": null,
        "name": "event-name",
        "priceCategory": null,
        "special": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`event mutations:  create event create an event with only required information 1`] = `
{
  "data": {
    "createEvent": {
      "event": {
        "admissionFee": null,
        "admissionFeeWithDiscount": null,
        "amountOfFloors": null,
        "club": {
          "id": 1,
          "name": "name",
        },
        "date": "event-date",
        "description": null,
        "genres": [],
        "id": 2,
        "imageUrls": [],
        "links": [],
        "minimumAge": null,
        "name": "event-name",
        "priceCategory": null,
        "special": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`event mutations:  create event create an event without permission shoul yield error 1`] = `
[
  [GraphQLError: You need to be logged in as admin to mutate data],
]
`;

exports[`event mutations:  delte event without permission should yield error 1`] = `
[
  [GraphQLError: You need to be logged in as admin to mutate data],
]
`;

exports[`event mutations:  update event to only have required fields 1`] = `
{
  "data": {
    "updateEvent": {
      "event": {
        "admissionFee": null,
        "admissionFeeWithDiscount": null,
        "amountOfFloors": null,
        "club": {
          "id": 2,
          "name": "name",
        },
        "date": "updated-date",
        "description": null,
        "genres": [],
        "id": 1,
        "imageUrls": [],
        "links": [],
        "minimumAge": null,
        "name": "updated-name",
        "priceCategory": null,
        "special": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`event mutations:  update event with all fields 1`] = `
{
  "data": {
    "updateEvent": {
      "event": {
        "admissionFee": 20,
        "admissionFeeWithDiscount": 3,
        "amountOfFloors": 2,
        "club": {
          "id": 2,
          "name": "name",
        },
        "date": "updated-date",
        "description": "updated-description",
        "genres": [
          {
            "id": 2,
            "name": "genre2",
          },
          {
            "id": 3,
            "name": "genre3",
          },
        ],
        "id": 1,
        "imageUrls": [
          "image-url-1",
          "image-url-2",
        ],
        "links": [
          {
            "href": "updated-link",
            "type": "FACEBOOK",
          },
        ],
        "minimumAge": 16,
        "name": "updated-name",
        "priceCategory": 1,
        "special": "updated-special",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`event mutations:  update event with emojis in description 1`] = `
{
  "data": {
    "updateEvent": {
      "event": {
        "admissionFee": null,
        "admissionFeeWithDiscount": null,
        "amountOfFloors": null,
        "club": {
          "id": 2,
          "name": "name",
        },
        "date": "updated-date",
        "description": "üçÉüîùüéâüì¢üëâüåµüéµ",
        "genres": [],
        "id": 1,
        "imageUrls": [],
        "links": [],
        "minimumAge": null,
        "name": "updated-name",
        "priceCategory": null,
        "special": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`event mutations:  update event with empty genre list 1`] = `
{
  "data": {
    "updateEvent": {
      "event": {
        "admissionFee": null,
        "admissionFeeWithDiscount": null,
        "amountOfFloors": null,
        "club": {
          "id": 2,
          "name": "name",
        },
        "date": "updated-date",
        "description": null,
        "genres": [],
        "id": 1,
        "imageUrls": [],
        "links": [],
        "minimumAge": null,
        "name": "updated-name",
        "priceCategory": null,
        "special": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`event mutations:  update event without permission should yield error 1`] = `
{
  "data": null,
  "errors": [
    [GraphQLError: You need to be logged in as admin to mutate data],
  ],
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;
